### PR TITLE
fix: skip flattening spread object with __proto__

### LINKED ISSUE #14756: [Bug]: JSX transform flattens spreads with __proto__
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

Programmatic API (`babel.transform`, `babel.parse`)

### Input code

```jsx
<div {...{__proto__: null}}/>
```

### Configuration file name

_No response_

### Configuration

```js
{
  plugins: ['@babel/plugin-transform-react-jsx'],
}
```

### Current and expected behavior

**current**
```js
"use strict";

/*#__PURE__*/
React.createElement("div", {
  __proto__: null
});
```

**expected**
```js
"use strict";

/*#__PURE__*/
React.createElement("div", _objectSpread({}, {
  __proto__: null
}));
```

### Environment

  System:
    OS: macOS 12.3
  Binaries:
    Node: 16.15.0 - ~/Library/Caches/fnm_multishells/52478_1657785102300/bin/node
    Yarn: 3.1.1 - /usr/local/bin/yarn
    npm: 8.5.5 - ~/Library/Caches/fnm_multishells/52478_1657785102300/bin/npm
  Monorepos:
    Yarn Workspaces: 3.1.1
  npmPackages:
    @babel/cli: ^7.17.10 => 7.17.10 
    @babel/core: ^7.18.0 => 7.18.0 
    @babel/eslint-config-internal: workspace:^ => 7.16.1 
    @babel/eslint-parser: workspace:^ => 7.18.2 
    @babel/eslint-plugin-development: workspace:^ => 7.17.7 
    @babel/eslint-plugin-development-internal: workspace:^ => 7.18.6 
    @babel/plugin-proposal-dynamic-import: ^7.16.7 => 7.16.7 
    @babel/plugin-proposal-export-namespace-from: ^7.17.12 => 7.17.12 
    @babel/plugin-proposal-object-rest-spread: ^7.18.0 => 7.18.0 
    @babel/plugin-transform-modules-commonjs: ^7.18.0 => 7.18.0 
    @babel/plugin-transform-runtime: ^7.18.0 => 7.18.0 
    @babel/preset-env: ^7.18.0 => 7.18.0 
    @babel/preset-flow: ^7.17.12 => 7.17.12 
    @babel/preset-typescript: ^7.17.12 => 7.17.12 
    @babel/runtime: ^7.18.0 => 7.18.0 
    babel-plugin-transform-charcodes: ^0.2.0 => 0.2.0 
    eslint: ^8.9.0 => 8.9.0 
    jest: ^28.0.3 => 28.0.3 

### Possible solution

_No response_

### Additional context

It has been suggested [here](https://github.com/microsoft/TypeScript/issues/49099#issuecomment-1126597811) that this is probably a bug in Babel.

### PR DESCRIPTION
**Pull request by JLHwung**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #14756
| Patch: Bug Fix?          | Y
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

In this PR we avoid optimizing the JSX pattern `{ ... { __proto__: foo }` to `{ __proto__: foo }` as JSX spread element evaluates to a plain object created by the global `Object` function.

This PR is a follow-up to #12557.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/14759"></a>

### CODE REVIEW

**Review by nicolo-ribaudo — APPROVED** [review]

File: packages/babel-plugin-transform-react-jsx/test/fixtures/react/handle-spread-with-proto/output.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+/*#__PURE__*/
+React.createElement("p", {
+  __proto__: null
+}, "text");
+
+/*#__PURE__*/
+React.createElement("div", {
+  "__proto__": null
+}, contents);
```
Comment (liuxingbaoyu) [inline packages/babel-plugin-transform-react-jsx/test/fixtures/react/handle-spread-with-proto/output.js:11]: Should `spreadElement` be used here?
It seems to be the expected behavior in https://github.com/babel/babel/issues/14756#issue-1304441362.

File: packages/babel-plugin-transform-react-jsx/src/create-plugin.ts [hunk 2]
```diff
@@ -422,7 +431,7 @@ You can set \`throwIfNamespace: false\` to bypass this warning.`,
       if (t.isJSXSpreadAttribute(attribute.node)) {
         const arg = attribute.node.argument;
         // Collect properties into props array if spreading object expression
-        if (t.isObjectExpression(arg)) {
+        if (t.isObjectExpression(arg) && !hasProto(arg)) {
```
Comment (liuxingbaoyu) [inline packages/babel-plugin-transform-react-jsx/src/create-plugin.ts:434]: `hasProto` seems to always return an array. Can we use `[].find`?

File: packages/babel-plugin-transform-react-jsx/test/fixtures/react/handle-spread-with-proto/output.js (hunk 1 above)
Comment (JLHwung) [reply packages/babel-plugin-transform-react-jsx/test/fixtures/react/handle-spread-with-proto/output.js:11]: Good catch. Yeah the spread element should not be stripped. I will investigate.

**Review by liuxingbaoyu — APPROVED** [review]

File: packages/babel-plugin-transform-react-jsx/test/fixtures/react/proto-in-jsx-attribute/input.js [hunk 3]
```diff
@@ -0,0 +1 @@
+<p __proto__={null} class="bar">text</p>;
```
Comment (JLHwung) [inline packages/babel-plugin-transform-react-jsx/test/fixtures/react/proto-in-jsx-attribute/input.js:1]: Both TS and Babel interpret `__proto__` in JSXAttributeKey as the special `__proto__` accessor, though JSX spec does not specify such behaviour.

File: packages/babel-plugin-transform-react-jsx/src/create-plugin.ts [hunk 4]
```diff
@@ -718,7 +727,17 @@ You can set \`throwIfNamespace: false\` to bypass this warning.`,
...
 
           if (objs.length === 1) {
-            return objs[0];
+            if (
+              !(
+                t.isSpreadElement(props[0]) &&
+                // If an object expression is spread element's argument
+                // it is very likely to contain __proto__ and we should stop
+                // optimizing spread element
+                t.isObjectExpression(props[0].argument)
```
Comment (JLHwung) [inline packages/babel-plugin-transform-react-jsx/src/create-plugin.ts:736]: The check here is necessary if we have pattern `{...{ __proto__ }}`, but it is not sufficient: `{...{...{ foo }}}`, after optimized into `{...{ foo }}` by `accumulateAttribute`, won't further pass the check here. I think this is fine as practically such pattern is rare.

File: packages/babel-plugin-transform-react-jsx/src/create-plugin.ts [hunk 5]
```diff
@@ -39,6 +39,15 @@ const get = (pass: PluginPass, name: string) =>
 const set = (pass: PluginPass, name: string, v: any) =>
   pass.set(`@babel/plugin-react-jsx/${name}`, v);
 
+function hasProto(node: t.ObjectExpression) {
+  return node.properties.some(
+    value =>
+      t.isObjectProperty(value, { computed: false }) &&
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-react-jsx/src/create-plugin.ts:45]:
```suggestion
      t.isObjectProperty(value, { computed: false, shorthand: false }) &&
```

Shorthand `__proto__` doesn't set the prototype.

**Comment by liuxingbaoyu** [comment]
This looks more like a react related PR.
:)

File: packages/babel-plugin-transform-react-jsx/test/fixtures/react/handle-spread-with-proto-babel-7/input.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+var __proto__ = null;
+
+<p {...{__proto__: null}}>text</p>;
+
+<div {...{"__proto__": null}}>{contents}</div>;
+
+<img alt="" {...{__proto__}} />;
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-react-jsx/test/fixtures/react/handle-spread-with-proto-babel-7/input.js:7]: This should behave like the computed fields, could you move it to `packages/babel-plugin-transform-react-jsx/test/fixtures/react/flattens-spread/input.js`?

**Comment by JLHwung** [comment]
I marked the PR as `jsx` because it changes how we transpile JSX. JSX is a syntax sugar providing the ability of representing call expressions with XML-like syntax. Though mostly used within React communities, JSX and React are decoupled. And that's why we have React alternatives like preact that also supports JSX.

I will mark a PR as `react` if it solves a React-related problems, e.g. mark `React.createElement` as pure so that it can be tree-shaken by bundlers.

**Review by nicolo-ribaudo — APPROVED** [review]

**Comment by ZipingL** [comment]
Hey Amazon Uses a lot of Babel. [Also Amazon has been have some IT issues with keeping track of my wages](https://jeffbezos.art), they stopped sending me like the right amounts, Is there where I ask for support when the DOLGOV suer as heck couldn't do much.

**Pull request merged**